include/zephyr/net/net_event.h

```diff
@@ -53,6 +54,20 @@ enum net_event_if_cmd {
 #define NET_EVENT_IF_ADMIN_UP				\
 	(_NET_EVENT_IF_BASE | NET_EVENT_IF_CMD_ADMIN_UP)
 
+/* Note to reviewers:
+ * This likely should be renamed if we keep NET_IF_OPT_CONNECTIVITY_MAX_TRIES
```

Comment:
@rlubos 

To me, persistent connectivity is connectivity that, at least with default settings, never gives up on being connected.

To be clear, `NET_IF_OPT_CONNECTIVITY_MAX_TRIES`, is only a consecutive counter. If connection succeeds, the total number of tries will be reset to zero. Some ifaces may not even have a concept of separate connection events, and on such ifaces `NET_IF_OPT_CONNECTIVITY_MAX_TRIES` could be ignored.

So, with an iface that is persistent, the  `NET_IF_OPT_CONNECTIVITY_MAX_TRIES` and `NET_IF_OPT_CONNECTIVITY_TIMEOUT` options basically specify special conditions under which even a **_persistent_** connection should give up (specifically when trying to transition from dormant to oper-up)

That means that:
> It's up to application to stop the cycle if persistence is enabled

Is only true if `NET_IF_OPT_CONNECTIVITY_MAX_TRIES` and `NET_IF_OPT_CONNECTIVITY_TIMEOUT` are both zero, at least as the RFC is currently written.

Having this built in **_may_** be a good idea just because it is a very common desire among actual applications -- But it also wouldn't be the end of the world if applications have to implement these behaviors themselves using events (in which case we should add an event specifically for "a connection attempt failed, but I am still trying to connect"). So really, it would be good to get input from others.

But that's just for persistent connectivity. What about non-persistent connectivity?

Conceivably, non-persistent connectivity could be defined two ways:

- Will try repeatedly to connect on the ***first*** transition from dormant to oper-up, but if connection is ever lost after that, gives up
- Will try only once, ever, to connect

Maybe the second definition is cleaner to adopt, since it allows the `MAX_TRIES` option to be for persistent networks only.

But even adopting that second definition, `CONNECTIVITY_TIMEOUT` still applies, and if it is infinite, even a non-persistent iface could wind up trying to connect forever anyways -- in which case, maybe `MAX_TRIES` ***should*** be an optional parameter for non-persistent ifaces.


I am neutral on both decisions. We just need to make a decision. To summarize:

1. Decide whether connection timeout and connection max_tries should be built-in, or up to the application to implement
2. If built-in, decide whether max_tries should be defined for non-persistent ifaces as well (and apply only during the first attempt to transition from dormant to oper-up)

Reply:
I think this feature is tricky because underlying protocols might differ how they approach the issue. I'd be even keen to park the persistence feature to gather more direct feedback about L2s more rather than make an API that isn't well aligned.

Having said that, I personally think that in this case it's a bit easier to deal with time than with retries. This is for two reasons:

1. In order to implement retries the underlying L2 must report connection retry attempts. This might not be available.
2. It might be easier to reason about power consumption. I could assume the worst case power consumption for connection attempts (CPU on, Radio On) and calculate energy expenditure. With retries I need to understand particular L2 behavior. 

I would prefer not to distinguish between first and consecutive connection attempts to make the API easier to grasp.

How about simplifying as follows:

* Keep `NET_EVENT_IF_CMD_CONNECTIVITY_TIMEOUT` opt and remove everything else.
* For the `NET_EVENT_IF_CMD_CONNECTIVITY_TIMEOUT` set to a non zero value it is expected that L2 will attempt to change from `dormant(true)` to `dormant(false)` state for at most `NET_EVENT_IF_CMD_CONNECTIVITY_TIMEOUT` seconds.
* For `NET_EVENT_IF_CMD_CONNECTIVITY_TIMEOUT` is set to 0 (?) the behavior is L2 implementation specific.

Reply:
You're right, introducing max_tries does require us to distinguish between two different types of connection attempt: The "macro" attempt, which may (or may not, depending on the L2) be a series of "micro" attempts. That adds confusion.

I'm on-board with the simplification you propose. It has the added benefit of not distinguishing between first and subsequent "macro" attempts "and also it doesn't distinguish between micro and macro connections at all", and I think that simplification is worth the minor loss of functionality.